### PR TITLE
Painting improvement: Added a palette component for spraycans and palette items.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
I have spent this day and the last one working on something I originally suggested to Anturk when he first introduced palette items, which is a component that manages a radial menu to more rapidly select colors from. I have also gone an extra mile and included a TGUI implementation of this feature, as I remember Anturk saying a TGUI color picker/palette would be better than just a radial menu in that same PR. More information can be found in the doc comments of the component itself, but it really isn't anything complicated, aside from the fact I had to tinker with the radial menu code to fit this particular one in.

## Why It's Good For The Game
Painting is fun but it has no such thing as a graphical palette to keep references of colors that are supposedly being used on the canvas, which makes any sufficiently serious attempt at pixel art (or reproducing a piece that's already been made outside the game for that matter) a hassle. This PR will improve this aspect hopefully, making it almost as good as a barebone pre-installed drawing software in terms of pixel art at the least.

Showcase:

https://user-images.githubusercontent.com/42542238/159139236-1208fc7f-9826-4ccf-bbfe-1e236b284531.mp4

## Changelog

:cl: Ghommie
add: Added a palette component for palettes and spraycans: Right click the item while it's in your active hand to access a radial menu that allows you to store, select or edit up to 14 colors for palettes and 8 for spraycans. Right-Click one of the color to pick a new color value, Left-Click to select it.
qol: Furthermore, a graphical palette containing the colors of the palette component will show up in the Canvas UI when painting with a tool that has said component, allowing you to select colors without unfocusing from the UI window.
del: Removed the ctrl-click (color picker) shortcut from spraycans, as I felt it to be too clunky and no longer necessary now (I'm the person that made it btw).
/:cl:

### CODE REVIEW

**Comment by Twaticus** [comment]
holy shit this is awesome

**Comment by bluetshirtguy** [comment]
amazing

**Comment by Cheshify** [comment]
Now let us refill spray cans so we can keep our palette

**Comment by Ghommie** [comment]
> Now let us refill spray cans so we can keep our palette

Ok, perhaps this may be a problem in the future, yet as of now, an empty and otherwise useless spraycan that has run out of charges can still be used for painting. This is a clear oversight, but until it isn't fixed for the better or worse, there's little reason for me to actually implement such, albeit nice, idea. Additionally, it's outside the scope of the PR.

**Comment by Cheshify** [comment]
> Ok, perhaps this may be a problem in the future, yet as of now, an empty and otherwise useless spraycan that has run out of charges can still be used for painting. This is a clear oversight, but until it isn't fixed for the better or worse, there's little reason for me to actually implement such, albeit nice, idea. Additionally, it's outside the scope of the PR.

Entirely based and fair, good PR by the way

**Comment by Farquaar** [comment]
> Now let us refill spray cans so we can keep our palette

Pretty sure spraycans only run out if you're painting the floor or walls. I've never had a spraycan run out on me while I'm painting a canvas, even if I'm painting multiple paintings with the same can.

**Comment by thestubborn** [comment]
> > Now let us refill spray cans so we can keep our palette
> 
> Ok, perhaps this may be a problem in the future, yet as of now, an empty and otherwise useless spraycan that has run out of charges can still be used for painting. This is a clear oversight, but until it isn't fixed for the better or worse, there's little reason for me to actually implement such, albeit nice, idea. Additionally, it's outside the scope of the PR.

i personally don't think that's an issue
if some assistant uses his bottomless spray can to make dickpaintings, delete the paintings and...paint...ban him (is that a thing?)

File: tgui/packages/tgui/interfaces/Canvas.tsx [hunk 1]
```diff
@@ -218,7 +226,26 @@ export const Canvas = (props, context) => {
             onCanvasModifiedHandler={(changed) => act("paint", { data: toMassPaintFormat(changed) })}
             editable={data.editable}
           />
-          <Flex align="center" justify="center">
+          <Flex align="center" justify="center" direction="column">
```
Comment (AnturK) [inline tgui/packages/tgui/interfaces/Canvas.tsx:229]: What's the purpose of this flex if you put it all in one element anyway

File: code/modules/art/paintings.dm [hunk 2]
```diff
@@ -149,6 +155,9 @@
 			used = TRUE
 			update_appearance()
 			. = TRUE
+		if("select_color")
+			var/obj/item/painting_implement = user.get_active_held_item()
+			painting_implement?.set_painting_tool_color(params["selected_color"])
```
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/art/paintings.dm:160]: This sort of validation can be already done in their version of the `set_painting_tool_color` proc. There should be no need to add one here or another proc for that.

File: tgui/packages/tgui/interfaces/Canvas.tsx (hunk 1 above)
Comment (AnturK) [reply tgui/packages/tgui/interfaces/Canvas.tsx:229]: Oh, sorry i misread that as separate tag.

**Pull request merged**